### PR TITLE
[Security] reusing AuthorizationChecker in AccessListener

### PR DESCRIPTION
**Pull request by oleg-andreyev**
| Q             | A
| ------------- | ---
| Branch?       | master
| Bug fix?      | yes
| New feature?  | no
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| License       | MIT
| Fixed tickets | #21571

While investigating why `AuthenticationEvents::AUTHENTICATION_SUCCESS` is not dispatched, noticed similarity between `AccessListener` and `AuthorizationChecker` and the only real difference is `alwaysAuthenticate` flag.

**Case 1:**
When `UsernamePasswordToken` is serialized it will be serialized with state of `authenticated`,
so when application will load token from session `AccessListener` won't call `authenticate` because `authenticated=true`, even if `alwaysAuthenticate=true`

**Case 2:**
User A has roles `foo, bar and admin`, User A is signed-in into application and token is persisted, later another User B with role `admin`, decided to restrict role `admin` for User A, so User A won't lose it's privileges until session is expired or logout, because token is persisted with `roles` and `authenticated=true` and AccessListener does not call `authenticate` even if `alwaysAuthenticate=true`

**TODO**
- [x] - reuse AuthorizationChecker in AccessListener
- [ ] - adjust unit tests
- [x] - add deprecation notices
- [x] - update CHANGELOG and UPGRADE*

### CODE REVIEW

File: src/Symfony/Component/Security/Http/Firewall/AccessListener.php
Comment (iltar) [inline src/Symfony/Component/Security/Http/Firewall/AccessListener.php]: This is a BC break, so it can't be merged as is. Not sure if BC breaks should be added in older versions or if we should not see this class as an extension point (which it is atm).

**Review by linaori — CHANGES_REQUESTED** [review]

File: src/Symfony/Component/Security/Http/Firewall/AccessListener.php
Comment (stof) [reply src/Symfony/Component/Security/Http/Firewall/AccessListener.php]: even then, we still need to preserve BC even in master

Comment (oleg-andreyev) [reply src/Symfony/Component/Security/Http/Firewall/AccessListener.php]: sure, I'll prepare BC changes in near time.

File: src/Symfony/Component/Security/Http/Firewall/AccessListener.php
Comment (iltar) [inline src/Symfony/Component/Security/Http/Firewall/AccessListener.php]: The docblocks and deprecation notices are not required for private properties (if the type is known). Private means it's not an extension point, thus don't have to marked as such.

File: src/Symfony/Component/Security/Http/Firewall/AccessListener.php
Comment (iltar) [inline src/Symfony/Component/Security/Http/Firewall/AccessListener.php]: This message should mentioned what's deprecated as of when and how to fix it

**Review by linaori — CHANGES_REQUESTED** [review]

File: src/Symfony/Component/Security/Http/Firewall/AccessListener.php
Comment (oleg-andreyev) [reply src/Symfony/Component/Security/Http/Firewall/AccessListener.php]: Updated deprecation message, but I would like to get suggestions on better wording

File: src/Symfony/Component/Security/Http/Firewall/AccessListener.php
Comment (oleg-andreyev) [reply src/Symfony/Component/Security/Http/Firewall/AccessListener.php]: removed docblock. thanks.

File: src/Symfony/Component/Security/Http/Firewall/AccessListener.php
Comment (oleg-andreyev) [inline src/Symfony/Component/Security/Http/Firewall/AccessListener.php]: `X.X` will be replaced with correspond version as soon as decision will be made where this code will land.

**Comment by oleg-andreyev** [comment]
@nicolas-grekas @iltar @stof can you give an advice against what branch this changes should go?
So 2.7 version entered security fixes phase 2 days ago.

**Comment by iltar** [comment]
My suggestion would be to fix this in 2.7 if it's a bug (which it does sound like), and then a feature for 4.2 to fix the constructor, making it an extra optional argument until the change of signature.

**Comment by oleg-andreyev** [comment]
Ok, I'll prepare my PR for 2.7 as bugfix, and then will prepare PR with deprecation for master

**Comment by oleg-andreyev** [comment]
@iltar @stof @nicolas-grekas all set and could be reviewed.

File: src/Symfony/Bundle/SecurityBundle/Resources/config/security_listeners.xml
Comment (iltar) [inline src/Symfony/Bundle/SecurityBundle/Resources/config/security_listeners.xml]: I'm not entirely sure the `on-invalid="null"` is needed. In case of an additional argument missing (old component version), it would just pass an extra argument to the class, which is allowed by php. So I think this can be a normal required argument.

File: src/Symfony/Component/Security/Http/Tests/Firewall/AccessListenerTest.php
Comment (iltar) [inline src/Symfony/Component/Security/Http/Tests/Firewall/AccessListenerTest.php]: Not sure if there's a standard defined for this, but you can also write:

```php
$this->event->method('getRequest')->willReturn($this->request);
```

File: src/Symfony/Component/Security/Http/Tests/Firewall/AccessListenerTest.php
Comment (iltar) [inline src/Symfony/Component/Security/Http/Tests/Firewall/AccessListenerTest.php]: Is this really needed? Can just set it to true I suppose. It looks like this logic was duplicated from the UsernamePasswordToken, but doesn't exactly serve a purpose. You can also be authenticated without roles.

**Review by linaori — CHANGES_REQUESTED** [review]

File: src/Symfony/Component/Security/Http/Tests/Firewall/AccessListenerTest.php
Comment (oleg-andreyev) [reply src/Symfony/Component/Security/Http/Tests/Firewall/AccessListenerTest.php]: Personally I do prefer shortcuts like `willReturn`, but I decided to stick with style that was before

File: src/Symfony/Component/Security/Http/Tests/Firewall/AccessListenerTest.php
Comment (oleg-andreyev) [reply src/Symfony/Component/Security/Http/Tests/Firewall/AccessListenerTest.php]: Indeed I wanted to mimic `UsernamePasswordToken`, but `UsernamePasswordToken` has a lot of parameters in it's constructor (lazy)

File: src/Symfony/Bundle/SecurityBundle/Resources/config/security_listeners.xml
Comment (oleg-andreyev) [reply src/Symfony/Bundle/SecurityBundle/Resources/config/security_listeners.xml]: You're right.

**Comment by iltar** [comment]
Have you tested what happens if "always authenticated" is set to true?

**Comment by oleg-andreyev** [comment]
Yes, that was the main reason why I reused `AuthorizationChecker`

**Comment by oleg-andreyev** [comment]
@nicolas-grekas any chances that this PR could be merged into 2.7?

File: src/Symfony/Component/Security/Http/Firewall/AccessListener.php
Comment (nicolas-grekas) [inline src/Symfony/Component/Security/Http/Firewall/AccessListener.php]: All the previous lines in the new methods duplicate the logic in "handle".
Would it make sense to not add any newmethod, and instead move the "if" block throwing "AuthenticationCredentialsNotFoundException" after these lines in "handle"?

File: src/Symfony/Component/Security/Http/Tests/Firewall/AccessListenerTest.php
Comment (nicolas-grekas) [inline src/Symfony/Component/Security/Http/Tests/Firewall/AccessListenerTest.php]: I'm not sure about this change. We prefer not doing refactos of the test suite. Could you revert these changes?

Comment (oleg-andreyev) [reply src/Symfony/Component/Security/Http/Tests/Firewall/AccessListenerTest.php]: No problem, I'll revert refactoring.

File: src/Symfony/Component/Security/Http/Firewall/AccessListener.php
Comment (nicolas-grekas) [inline src/Symfony/Component/Security/Http/Firewall/AccessListener.php]: shouldn't this be "if (!...isGranted()) throw..."?

Comment (oleg-andreyev) [reply src/Symfony/Component/Security/Http/Firewall/AccessListener.php]: True, I've adjusted the code

**Comment by nicolas-grekas** [comment]
LGTM code-style wise. But since I'm not an expert in this component, I'll let others approve the PR.

**Comment by oleg-andreyev** [comment]
@nicolas-grekas who has more experience in security component? Got an email with reminder about 2.7 that it won't be maintained expect security fixes

**Review by wouterj — APPROVED** [review]
Looks like a very valid bug fix to me. It makes sense to de-duplicate logic and the described scenario's do appear to be the correct behaviour.

`always_authenticate_before_granting` is disabled by default, so this will only impact projects that explicitly wanted the described behaviour.

:+1: for merging this in 2.7

**Review by dunglas — APPROVED** [review]

**Comment by fabpot** [comment]
This is more a behavior change than a bug fix to me. I don't think this is something that can be merged in 2.7.

**Comment by chalasr** [comment]
Agreed. The bug is in the current documentation, not in the code.

**Comment by oleg-andreyev** [comment]
I agree that it's a behavior change, but to me current behavior is unexpected (and not documented) which is caused by inconsistent logic between AccessListener and AuthorizationChecker. 

I usually treat unexpected behavior as bug.

But if majority agrees, I'll rebase changes on master all together will deprecations for AccessListener

**Comment by fabpot** [comment]
Drawing the line between a bug and a behavior change is always very difficult. But in this case, I think this is clearly not something we can treat as a bug (especially case 2 which has been reported several times and even if this looks "better" to have the new behavior, this is really a new feature).

**Comment by wouterj** [comment]
The inconsistency bit here is what makes this PR very difficult to mark as feature or bug.

For instance:

**Example 1.1**
```yaml
security:
    # ...
    access_control:
        - { pattern: ^/admin, roles: 'ROLE_ADMIN' }
```

**Example 1.2**
```php
/** @Route("/admin") */
public function admin()
{
    $this->denyAccessUnlessGranted('ROLE_ADMIN');
}
```

Before this PR, these two would act differently: (1.1) will not re-authenticate when `alwaysAuthenticated` is true, while (1.2) will.

---

Furthermore, scenario 2 described in the PR description will only act in this way if `always_authenticate` is explicitly set to `true`. A default Symfony app (i.e. most of the app around) would not have different behaviour.

All current authentication managers and providers are already capable of dealing with re-authentication, because example (1.2) is a feature since the very start of Symfony.

---

Given these three points, I think this is not so much a behaviour change as it is about fixing inconsistent behaviour. That's why I would indicate this as a bug fix.

---

Anyway, the core team is set-up to precisily define whether something is a bug or feature. So feel free to merge in whatever branch you like :smiley:

**Comment by oleg-andreyev** [comment]
Changed base to 3.4.

File: src/Symfony/Component/Security/Http/Firewall/AccessListener.php
Comment (nicolas-grekas) [inline src/Symfony/Component/Security/Http/Firewall/AccessListener.php]: I think we should keep this block. The exception message is more tailored than the one in `AuthorizationChecker`

**Review by nicolas-grekas — CHANGES_REQUESTED** [review]
This is staling because changing the behavior on lower branches poses a risk to break BC for someone.
Let's merge on master. If there is anything to warn about in UPGRADE* file and CHANGELOG of the component, it should be added also.
OK to move forward with this plan @oleg-andreyev?

File: src/Symfony/Component/Security/Http/Firewall/AccessListener.php
Comment (oleg-andreyev) [reply src/Symfony/Component/Security/Http/Firewall/AccessListener.php]: In my opinion message from `AuthorizationChecker` is better from DX perspective, because it is hinting about why token may be missing.

Comment (nicolas-grekas) [reply src/Symfony/Component/Security/Http/Firewall/AccessListener.php]: The message in `AuthorizationChecker` reads as

> The token storage contains no authentication token. One possible reason may be that there is no firewall configured for this URL.

Is it possible to have the AccessListener configured but not firewall configured? If not, then the message would be misleading.

But there is something else: this moves the check for a token after the check for attributes. Is it legitimate?

Comment (oleg-andreyev) [reply src/Symfony/Component/Security/Http/Firewall/AccessListener.php]: `AccessListener` is configured by https://github.com/symfony/symfony/blob/5e28ac31504e3a7cd2d16b0216182d767f37ad41/src/Symfony/Bundle/SecurityBundle/DependencyInjection/SecurityExtension.php#L169

So it's impossible to have configured `AccessListener` without firewalls, unless developers are doing it somehow else.

Comment (nicolas-grekas) [reply src/Symfony/Component/Security/Http/Firewall/AccessListener.php]: OK to me, lets continue here :)

**Comment by xabbuh** [comment]
If some of the proposed changes are a bugfix, we need to split this PR: Move everything that is not about the bugfix to a new PR targetting `master` and then rebase the changes for the bugfix onto the `3.4` branch.

**Comment by oleg-andreyev** [comment]
This PR got messy, I'm closing it and will provide a cleaner version.

**Pull request closed without merging**
